### PR TITLE
✨ Add TokenHistory To Boundary Events

### PR DESCRIPTION
**Pull request by SteffenKn**
> Related to https://github.com/process-engine/process_engine_runtime/pull/236

**Changes:**

1. Add TokenHistory to boundary event

**Issues:**

Related to https://github.com/process-engine/bpmn-studio/issues/1198

PR: #220

## How can others test the changes?

- **Have a look at https://github.com/process-engine/bpmn-studio/pull/1332**

## PR-Checklist

Please check the boxes in this list after submitting your PR:

- [x] You can merge this PR **right now** (if not, please prefix the title with "WIP: ")
- [x] I've tested **all** changes included in this PR.
- [x] I've also reviewed this PR myself before submitting (e.g. for scrambled letters, typos, etc.).
- [x] I've rebased the `develop` branch with my branch before finishing this PR.
- [x] I've **summarized all changes** in a list above.
- [x] I've mentioned all **PRs, which relate to this one**.
- [x] I've prefixed my Pull Request title is according to [gitmoji guide](https://gitmoji.carloscuesta.me/).

### CODE REVIEW

File: src/runtime/engine/flow_node_handler/boundary_event_handlers/error_boundary_event_handler.ts [hunk 1]
```diff
@@ -35,10 +35,17 @@ export class ErrorBoundaryEventHandler extends FlowNodeHandlerInterruptible<Mode
...
+
+    await this.persistOnEnter(token);
+
     try {
+      await this.persistOnExit(token);
+
       // Must use return await here to prevent unhandled rejections.
       return await this._decoratedHandler.execute(token, processTokenFacade, processModelFacade, identity, this.previousFlowNodeInstanceId);
     } catch (err) {
+      await this.persistOnError(token, err);
```
Comment (ElRaptorus) [inline src/runtime/engine/flow_node_handler/boundary_event_handlers/error_boundary_event_handler.ts]: Should be `await this.persistOnExit(token, err);` as well, because intercepting errors is the normal function of this BoundaryEvent.

The decorated handler will already have been put in an `error` state. You can use that one to illustrate an error in the LET.

File: src/runtime/engine/flow_node_handler/boundary_event_handlers/error_boundary_event_handler.ts [hunk 2]
```diff
@@ -49,11 +56,17 @@ export class ErrorBoundaryEventHandler extends FlowNodeHandlerInterruptible<Mode
...
+    await this.persistOnEnter(onEnterToken);
+
     try {
+      await this.persistOnExit(onEnterToken);
+
       // Must use return await here to prevent unhandled rejections.
       return await this._decoratedHandler.resume(flowNodeInstance, processTokenFacade, processModelFacade, identity);
     } catch (err) {
-      const onEnterToken: Runtime.Types.ProcessToken = flowNodeInstance.getTokenByType(Runtime.Types.ProcessTokenType.onEnter);
+      await this.persistOnError(onEnterToken, err);
```
Comment (ElRaptorus) [inline src/runtime/engine/flow_node_handler/boundary_event_handlers/error_boundary_event_handler.ts]: Like above, should be `await this.persistOnExit(token, err);` here as well.

File: src/runtime/engine/flow_node_handler/boundary_event_handlers/message_boundary_event_handler.ts [hunk 3]
```diff
@@ -68,6 +70,8 @@ export class MessageBoundaryEventHandler extends FlowNodeHandlerInterruptible<Mo
       this.handlerHasFinished = true;
 
       if (this.messageReceived) {
+        await this.persistOnExit(token);
```
Comment (ElRaptorus) [inline src/runtime/engine/flow_node_handler/boundary_event_handlers/message_boundary_event_handler.ts]: This state change must happen after this if-clause.

Persisting the State change upon message-recieval is already done in line [#138](https://github.com/process-engine/process_engine_core/pull/220/files#diff-d37eb0e0f9a6e4ec9b6d302f110aa4ebR138), so doing it here again would duplicate the onExit 

But we must also change the state of the BoundaryEvent, if the handler finishes before the message arrives.

So this must be: 

```ts
if (this.messageReceived) {
  return;
}

await this.persistOnExit(token);
```

File: src/runtime/engine/flow_node_handler/boundary_event_handlers/message_boundary_event_handler.ts [hunk 4]
```diff
@@ -99,6 +105,8 @@ export class MessageBoundaryEventHandler extends FlowNodeHandlerInterruptible<Mo
       this.handlerHasFinished = true;
 
       if (this.messageReceived) {
+        await this.persistOnExit(onEnterToken);
+
```
Comment (ElRaptorus) [inline src/runtime/engine/flow_node_handler/boundary_event_handlers/message_boundary_event_handler.ts]: Like above, this should be:

```ts
if (this.messageReceived) {
  return;
}

await this.persistOnExit(token);
```

File: src/runtime/engine/flow_node_handler/boundary_event_handlers/message_boundary_event_handler.ts [hunk 5]
```diff
@@ -124,9 +132,17 @@ export class MessageBoundaryEventHandler extends FlowNodeHandlerInterruptible<Mo
 
       this.messageReceived = true;
 
+      const previousFlowNodeInstanceId: string = token.flowNodeInstanceId;
+      token.flowNodeInstanceId = this.flowNodeInstanceId;
```
Comment (ElRaptorus) [inline src/runtime/engine/flow_node_handler/boundary_event_handlers/message_boundary_event_handler.ts]: I'm not sure I get why this is necessary?

File: src/runtime/engine/flow_node_handler/boundary_event_handlers/signal_boundary_event_handler.ts [hunk 6]
```diff
@@ -68,6 +69,8 @@ export class SignalBoundaryEventHandler extends FlowNodeHandlerInterruptible<Mod
       this.handlerHasFinished = true;
 
       if (this.signalReceived) {
+        await this.persistOnExit(token);
+
         return;
```
Comment (ElRaptorus) [inline src/runtime/engine/flow_node_handler/boundary_event_handlers/signal_boundary_event_handler.ts]: Like above, should be:

```ts
if (this.signalReceived) {
  return;
}

await this.persistOnExit(token);
```

File: src/runtime/engine/flow_node_handler/boundary_event_handlers/signal_boundary_event_handler.ts [hunk 7]
```diff
@@ -99,6 +104,8 @@ export class SignalBoundaryEventHandler extends FlowNodeHandlerInterruptible<Mod
       this.handlerHasFinished = true;
 
       if (this.signalReceived) {
+        await this.persistOnExit(onEnterToken);
```
Comment (ElRaptorus) [inline src/runtime/engine/flow_node_handler/boundary_event_handlers/signal_boundary_event_handler.ts]: Like above, should be:

```ts
if (this.signalReceived) {
  return;
}

await this.persistOnExit(token);
```

File: src/runtime/engine/flow_node_handler/boundary_event_handlers/signal_boundary_event_handler.ts [hunk 8]
```diff
@@ -129,6 +136,13 @@ export class SignalBoundaryEventHandler extends FlowNodeHandlerInterruptible<Mod
 
       await this._decoratedHandler.interrupt(token);
 
+      const previousFlowNodeInstanceId: string = token.flowNodeInstanceId;
+      token.flowNodeInstanceId = this.flowNodeInstanceId;
```
Comment (ElRaptorus) [inline src/runtime/engine/flow_node_handler/boundary_event_handlers/signal_boundary_event_handler.ts]: See above, why is this necessary?

File: src/runtime/engine/flow_node_handler/boundary_event_handlers/timer_boundary_event_handler.ts [hunk 9]
```diff
@@ -69,6 +72,8 @@ export class TimerBoundaryEventHandler extends FlowNodeHandlerInterruptible<Mode
       this.hasHandlerFinished = true;
 
       if (this.timerHasElapsed) {
+        await this.persistOnExit(token);
```
Comment (ElRaptorus) [inline src/runtime/engine/flow_node_handler/boundary_event_handlers/timer_boundary_event_handler.ts]: Like above, should be:

```ts
if (this.timerHasElapsed) {
  return;
}

await this.persistOnExit(token);
```

File: src/runtime/engine/flow_node_handler/boundary_event_handlers/timer_boundary_event_handler.ts [hunk 10]
```diff
@@ -93,12 +98,16 @@ export class TimerBoundaryEventHandler extends FlowNodeHandlerInterruptible<Mode
...
 
+      await this.persistOnEnter(onEnterToken);
+
       const nextFlowNodeInfo: NextFlowNodeInfo =
         await this._decoratedHandler.resume(flowNodeInstance, processTokenFacade, processModelFacade, identity);
 
       this.hasHandlerFinished = true;
 
       if (this.timerHasElapsed) {
+        await this.persistOnExit(onEnterToken);
```
Comment (ElRaptorus) [inline src/runtime/engine/flow_node_handler/boundary_event_handlers/timer_boundary_event_handler.ts]: Like above, should be:

```ts
if (this.timerHasElapsed) {
  return;
}

await this.persistOnExit(token);
```

File: src/runtime/engine/flow_node_handler/boundary_event_handlers/timer_boundary_event_handler.ts [hunk 11]
```diff
@@ -132,8 +141,16 @@ export class TimerBoundaryEventHandler extends FlowNodeHandlerInterruptible<Mode
       if (this.hasHandlerFinished) {
         return;
       }
+
       this.timerHasElapsed = true;
 
+      const previousFlowNodeInstanceId: string = token.flowNodeInstanceId;
+      token.flowNodeInstanceId = this.flowNodeInstanceId;
```
Comment (ElRaptorus) [inline src/runtime/engine/flow_node_handler/boundary_event_handlers/timer_boundary_event_handler.ts]: See above.

Comment (SteffenKn) [reply src/runtime/engine/flow_node_handler/boundary_event_handlers/timer_boundary_event_handler.ts]: We discussed this. This is necessary because otherwise the onExit token will not be added to the BoundaryEvent, but to the previous FlowNode.

Comment (ElRaptorus) [reply src/runtime/engine/flow_node_handler/boundary_event_handlers/timer_boundary_event_handler.ts]: I remember. But I'm not so sure that this is the right place to override the FlowNodeInstanceId, since all handlers use a common base handler. 

If an override is necessary, it should happen there, not in the derived handlers. Otherwise we'll just risk passing down the same error through all the other handlers.

**Comment by ElRaptorus** [comment]
@SteffenKn I hate to tell you this (well, not really 😜), but this has become obsolete through this PR:
https://github.com/process-engine/process_engine_core/pull/239

However, what is still missing is an `interrupted` state for FlowNodeInstances. 
You can still to that, if you like. 🙂
